homeassistant/helpers/template.py

```diff
@@ -1996,13 +2020,16 @@ class TemplateEnvironment(ImmutableSandboxedEnvironment):
 
     def __init__(self, hass, limited=False, strict=False):
         """Initialise template environment."""
+        undefined: type[LoggingUndefined] | type[jinja2.StrictUndefined]
         if not strict:
             undefined = LoggingUndefined
         else:
             undefined = jinja2.StrictUndefined
         super().__init__(undefined=undefined)
         self.hass = hass
-        self.template_cache = weakref.WeakValueDictionary()
+        self.template_cache: weakref.WeakValueDictionary[
+            str | jinja2.nodes.Template, CodeType | str | None
+        ] = weakref.WeakValueDictionary()
```

Comment:
Unrelated to this PR, but I wonder if we have duplicates here.

Reply:
Entirely possible, the typing is a bit odd (but accurate).